File: elixir/meetup/meetup.exs
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @weekday_map %{monday: 1, tuesday: 2, wednesday: 3, thursday: 4,
                 friday: 5, saturday: 6, sunday: 7}

  @type weekday ::
      :monday | :tuesday | :wednesday
    | :thursday | :friday | :saturday | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: :calendar.date
  def meetup(year, month, weekday, schedule) do
    {year, month, find_date(year, month, Map.get(@weekday_map, weekday), schedule)}
  end

  defp find_date(year, month, weekday, :first) do
    find_date(year, month, weekday, 1)
  end
  defp find_date(year, month, weekday, :second) do
    find_date(year, month, weekday, 8)
  end
  defp find_date(year, month, weekday, :third) do
    find_date(year, month, weekday, 15)
  end
  defp find_date(year, month, weekday, :fourth) do
    find_date(year, month, weekday, 22)
  end
  defp find_date(year, month, weekday, :teenth) do
    find_date(year, month, weekday, 13)
  end
  defp find_date(year, month, weekday, :last) do
    last_day = Enum.drop_while((31..28), fn(i) ->
      not :calendar.valid_date(year, month, i)
    end)
    |> Enum.take(1)
    |> hd
    Enum.drop_while((last_day..(last_day-7)), fn(i) ->
      :calendar.day_of_the_week(year, month, i) != weekday
    end)
    |> Enum.take(1)
    |> hd
  end

  defp find_date(year, month, weekday, start) do
    Enum.drop_while((start..start+7), fn(i) ->
      :calendar.day_of_the_week(year, month, i) != weekday
    end)
    |> Enum.take(1)
    |> hd
  end
end

File: clojure/meetup/src/meetup.clj
(ns meetup
  (import java.util.Calendar))

(def days {:sunday Calendar/SUNDAY
           :monday Calendar/MONDAY
           :tuesday Calendar/TUESDAY
           :wednesday Calendar/WEDNESDAY
           :thursday Calendar/THURSDAY
           :friday Calendar/FRIDAY
           :saturday Calendar/SATURDAY})

;; for each week name, value is [start-day, nth-day, day inc/dec]
(def week-ops {:first  [ 1 0  1]
               :second [ 1 1  1]
               :third  [ 1 2  1]
               :fourth [ 1 3  1]
               :last   [ 1 0 -1]
               :teenth [13 0  1]})

(defn meetup
  [month year day-name week]
  (let [d (days day-name)
        [start-day nth-day incdec-days] (week-ops week)
        c (Calendar/getInstance)]
    (.set c year (dec month) start-day 0 0 0)
    (when (= week :last)
      (.set c Calendar/DAY_OF_MONTH (.getActualMaximum c Calendar/DAY_OF_MONTH)))

    (loop [nth-day nth-day]
      (let [right-day-of-week? (= (.get c Calendar/DAY_OF_WEEK) d)]
      (cond (and right-day-of-week? (zero? nth-day))
            [year month (.get c Calendar/DAY_OF_MONTH)]

            right-day-of-week?
            (do (.add c Calendar/DAY_OF_MONTH (* 7 incdec-days))
                (recur (dec nth-day)))

            :else
            (do (.add c Calendar/DAY_OF_MONTH incdec-days)
                (recur nth-day)))))))
